django/utils/tree.py

```diff
@@ -47,9 +43,14 @@ def __str__(self):
     def __repr__(self):
         return "<%s: %s>" % (self.__class__.__name__, self)
 
+    def __copy__(self):
+        return self.create(self.children, self.connector, self.negated)
+
+    copy = __copy__
+    clone = __copy__
```

Comment:
Alas, we get a single test failure, confusingly in a test for `__deepcopy__`:

```python
======================================================================
FAIL: test_deepcopy (utils_tests.test_tree.NodeTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/utils_tests/test_tree.py", line 76, in test_deepcopy
    self.assertIs(self.node1.children, node4.children)
AssertionError: [('a', 1), ('b', 2)] is not [('a', 1), ('b', 2)]
```

This is actually a check for shallow copy, but our `__copy__` here is using `.create()` which calls `children[:]`. This is fixed by rolling back to the following:

```diff
diff --git a/django/utils/tree.py b/django/utils/tree.py
index 67b47f027c..8bc01d85fb 100644
--- a/django/utils/tree.py
+++ b/django/utils/tree.py
@@ -44,10 +44,14 @@ def __repr__(self):
         return "<%s: %s>" % (self.__class__.__name__, self)
 
     def __copy__(self):
-        return self.create(self.children, self.connector, self.negated)
+        obj = self.create()
+        obj.__dict__ = self.__dict__.copy()
+        return obj
 
     copy = __copy__
-    clone = __copy__
+
+    def clone(self):
+        return self.create(self.children, self.connector, self.negated)
 
     def __deepcopy__(self, memodict):
         obj = self.create(connector=self.connector, negated=self.negated)
```

So we have two options:

1. Keep it as is and don't worry that the copy is not as shallow... 😬 
2. Apply this patch.

If we go for option 2 we probably ought to check other uses of `.copy()` to decide whether we need to be using `.clone()` instead, notably in `.add()`, or we need to determine whether `.clone()` is really meant to be a `__deepcopy__` alias.

I don't really understand in a structure like this why we don't just have shallow and deep copying, but have this funny extra thing to clone one level down. Surely if we nest these objects then we want to deep copy to avoid any issues with changes deeper inside the tree?

@kezabelle Would you be able to run some performance checks over this? I'm not sure it'll necessarily be as good as I first had, but it shouldn't be worse than the original. I think that the tidy up to have a consistent approach here is most important, however, as it was getting crufty.

Reply:
Will do, marking it as unread to remind me.

Reply:
Ok, so I've reverted `__copy__()` to something similar in layout to `__deepcopy__()`:

```python
    def __copy__(self):
        obj = self.create(connector=self.connector, negated=self.negated)
        obj.children = self.children  # Don't [:] as .__init__() via .create() does.
        return obj
```